ee/app/livechat-enterprise/server/hooks/afterReturnRoomAsInquiry.ts

```diff
@@ -1,19 +1,44 @@
+import moment from 'moment';
+
 import { callbacks } from '../../../../../app/callbacks/server';
-import { LivechatRooms } from '../../../../../app/models/server';
+import { LivechatInquiry, LivechatRooms } from '../../../../../app/models/server';
 import { settings } from '../../../../../app/settings/server';
 
-const afterReturnRoomAsInquiry = ({ room }: { room: any }): void => {
+const unsetPredictedVisitorAbandonment = ({ room }: { room: any }): void => {
 	if (!room?._id || !room?.omnichannel?.predictedVisitorAbandonmentAt) {
 		return;
 	}
 
 	(LivechatRooms as any).unsetPredictedVisitorAbandonmentByRoomId(room._id);
 };
 
+const setQueueTimer = (timeToAdd: any) => ({ room }: { room: any }): void => {
+	if (!room?._id || room?.t !== 'l') {
+		return;
+	}
+
+	const inquiry = (LivechatInquiry as any).findOneByRoomId(room?._id);
+	if (!inquiry) {
+		return;
+	}
+
+	const newQueueTime = moment(inquiry?._updatedAt).add(timeToAdd, 'minutes');
+	(LivechatInquiry as any).setEstimatedInactivityCloseTime(inquiry?._id, newQueueTime);
+};
+
 settings.get('Livechat_abandoned_rooms_action', (_, value) => {
 	if (!value || value === 'none') {
 		callbacks.remove('livechat:afterReturnRoomAsInquiry', 'livechat-after-return-room-as-inquiry');
 		return;
 	}
-	callbacks.add('livechat:afterReturnRoomAsInquiry', afterReturnRoomAsInquiry, callbacks.priority.HIGH, 'livechat-after-return-room-as-inquiry');
+	callbacks.add('livechat:afterReturnRoomAsInquiry', unsetPredictedVisitorAbandonment, callbacks.priority.HIGH, 'livechat-after-return-room-as-inquiry');
+});
+
+settings.get('Livechat_max_queue_wait_time_action', (_, value) => {
+	if (!value || value === 'Nothing') {
+		callbacks.remove('livechat:afterReturnRoomAsInquiry', 'livechat-after-return-room-as-inquiry-set-queue-timer');
+		return;
+	}
+	const timer = settings.get('Livechat_max_queue_wait_time');
+	callbacks.add('livechat:afterReturnRoomAsInquiry', setQueueTimer(timer), callbacks.priority.HIGH, 'livechat-after-return-room-as-inquiry-set-queue-timer');
```

Comment:
We can't rely only on this callback to update the `estimated Inactivity close time` because this callback is only called when the routing system algorithm is `Manual Selection`, but Inquiries are queued in all algorithms, so we need to rely on a more consist solution that will cover all use cases.

So, we need to track this method -> `LivechatInquiry.queueInquiry`. As far as I know, the method is called twice in the code, so rather than calling `LivechatInquiry.queueInquiry`, we could maybe create a new method that would be called and inside the new method you'd call -> `LivechatInquiry.queueInquiry` and, also, run a new callback method, so rather than implementing this logic here, you could implement the logic for the new callback I just mentioned, something like:  `livechat:afterQueueInquiry`, WDYT?

You could, even, improve the signature of the method `Livechat.queueInquiry` and support additional data, so you could also pass along the new value of `estimatedInactivityCloseTimeAt` and update the document at once.

Reply:
Hey but this callback is called only on `returnRoomAsInquiry` method, this callback should just set the value again when the user returns the chat to the queue, and as far as i can see on `app/livechat/server/lib/Livechat.js L663` this should happen disregards of the routing algorithm.

The callback that sets the value when the chat first comes is on the `createLivechatInquiry` method (using the `livechat.beforeInquiry` cb) and that should also run on all routing algos, since it happens before queueing the inquiry.

Reply:
Regarding `returnRoomAsInquiry`:
- The method is called when a user returns the chat to the queue manually(Only available on Manual Selection)
- The method is called when the agent didn't reply and the chat is auto-transferred, however, there's a condition: `RoutingManager.getConfig().autoAssignAgent` ==== `Manual Selection`.

Apart from that, the chat can be queued in other places, such as here:

https://github.com/RocketChat/Rocket.Chat/blob/6267bf4b6d726f7a3709c1a7ed79f7ed4243d44c/app/livechat/server/lib/Helper.js#L409-L412

So, as you can see, the chat will be queued, but the `estimatedInactivityCloseTimeAt` won't be set, that's the reason we should implement a more centralized solution that would cover all cases and would reduce complexity.

Reply:
That's interesting to know. What about using the `livechat.beforeRouteChat` to set that value then? (or create another callback def at the same place as this) to avoid having to modify multiple files by changing the function signature?

Reply:
IMHO we should wrap the method `LivechatInquiry.queueInquiry`(called twice in the codebase) in a new method that, as the last step, would call a new callback, perhaps -> `livechat:afterQueueInquired` and, then, implement the logic for the new callback.
In this case, we wouldn't need to implement callbacks for `livechat.beforeInquiry` nor `livechat:afterReturnRoomAsInquiry`, just one single callback method would do the trick.